RFC: drivers: serial: Use microseconds to represent timeout

nordic-krch opened the pull request:
Updated uart_rx_enable() and uart_tx() to use timeout given
in microseconds. Previously argument was given in milliseconds.
However, there are cases when milliseconds granularity is not
enough and can significantly reduce a throughput, e.g. 1ms is
100 bytes at 1Mb.
    
Updated 4 drivers which implement asynchronous API. Updated
places where API was used.

Additionally, added `SYS_FOREVER_US` to represent inifinity in microseconds (similar to `SYS_FOREVER_MS`).

Fixes #30861.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
sys: time_units: Add SYS_FOREVER_US macro

Add macro for representing forever timeout in microseconds.
Macro is similar to already existing SYS_FOREVER_MS.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: Use microseconds to represent timeout

Updated uart_rx_enable() and uart_tx() to use timeout given
in microseconds. Previously argument was given in milliseconds.
However, there are cases when milliseconds granularity is not
enough and can significantly reduce a throughput, e.g. 1ms is
100 bytes at 1Mb.

Updated 4 drivers which implement asynchronous API. Updated
places where API was used.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

Mierunski commented:
Looks good, just wondering if we should switch to `int64_t` in tx/rx_timeout

dcpleung reviewed (approved):
LGTM. Since this is changing a stable API, this needs to go through the stable API change process.

nordic-krch commented:
@dcpleung According to this table https://docs.zephyrproject.org/latest/reference/api/overview.html it is still unstable.
@carlescufi Can you add this to API meeting agenda?

nordic-krch commented:
> wondering if we should switch to int64_t in tx/rx_timeout

It's >4200 seconds. To me it sounds enough but can be changed to 64 bits if there are any concerns.

dcpleung commented:
Ah... the UART async API is still unstable, but the UART API is. So not an issue then.

cfriedt commented:
@nordic-krch Can this be merged or is it still an RFC?

nordic-krch commented:
@cfriedt let's discuss it on api meeting today.

carlescufi commented on include/drivers/uart.h:
```diff
@@ -486,8 +486,8 @@ static inline int uart_callback_set(const struct device *dev,
  * @param dev     UART device structure.
  * @param buf     Pointer to transmit buffer.
  * @param len     Length of transmit buffer.
- * @param timeout Timeout in milliseconds. Valid only if flow control is
```
Open question: should we use `int64_t` to allow for longer timeouts?

anangl replied:
I think there is no need to. With `int32_t`, the maximum value gives almost 36 minutes, so it seems to be enough for UART communication.

nordic-krch committed:
doc: releases: Add note about UART Asynchronous API change

Added note to release notes about timeout units used in
uart_tx and uart_rx_enable.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/serial/uart_stm32.c:
```diff
@@ -808,10 +808,10 @@ static inline void async_evt_rx_buf_release(struct uart_stm32_data *data)
-	if ((timeout != SYS_FOREVER_MS) && (timeout != 0)) {
+	if ((timeout != SYS_FOREVER_US) && (timeout != 0)) {
 		/* start timer */
 		LOG_DBG("async timer started for %d ms", timeout);
```
```suggestion
		LOG_DBG("async timer started for %d us", timeout);
```

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1044,11 +1044,11 @@ static void rxstarted_isr(const struct device *dev)
-			      K_MSEC(data->async->rx_timeout_slab),
-			      K_MSEC(data->async->rx_timeout_slab));
+			      K_USEC(data->async->rx_timeout_slab),
+			      K_USEC(data->async->rx_timeout_slab));
```
Nit: two comments, in lines [929](https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L929) and [999](https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L999), still mention "ms" as the RX timeout unit.

Also the calculation of `rx_timeout_slab`:
https://github.com/zephyrproject-rtos/zephyr/blob/9208b14fde0e71b9c5bd9237c7e9af88f18a6022/drivers/serial/uart_nrfx_uarte.c#L809-L811
 should be updated, as the minimum value there is set for milliseconds.
